Added container error handling in PreRun/PostRun (#96)
NearNodeFlash/NearNodeFlash.github.io: docs/guides/user-containers/readme.md

    GFS2 filesystems have special considerations since the mount directory contains directories for
    every compute node. See [GFS2 Index Mounts](#gfs2-index-mounts) for more info.

### Targeting Nodes

For container directives, compute nodes must be assigned to the workflow. The NNF software will
trace the compute nodes back to their local NNF nodes and the containers will be executed on those
NNF nodes. The act of assigning compute nodes to your container workflow instructs the NNF software

> to select the NNF nodes that run the containers.

Reviewer: This last sentence seems redundant with the prior sentence.

For the `jobdw` directive that is included above, the servers (i.e. NNF nodes) must also be assigned
along with the computes.

> 

Reviewer: If a compute node not associated with any of the Rabbits assigned in the Servers object are specified, what happens?

Author: Good question. The NNF file system wouldn't be available on the rabbit and therefore the mount directory passed into the container won't exist. I'll look into this more and add any fixes, if necessary.

Reviewer: Maybe the webhook should check this?

## Running a Container Workflow

> Once the workflow is created, it will need to progress through the following states. This is a quick

Reviewer: ```suggestion
Once the workflow is created, the WLM progresses it through the following states. This is a quick
```

overview of the container-related behavior that occurs:

> - Proposal: Verify storages are provided according to the container profile.

Reviewer 2: Can "storages" contain a hyperlink up to the "Container Storages" section?  I still stumble on some of the short-hand lingo.

Author: Good call

- Setup: If applicable, request ports from NnfPortManager.
- DataIn: No container related activity.
- PreRun: Appropriate `MPIJob` or `Job(s)` are created for the workflow. In turn, user containers

> are created and launched by Kubernetes. Containers are expected to have started in this state.

Reviewer: ```suggestion
are created and launched by Kubernetes. Containers are expected to start in this state.
```

> - PostRun: Once in PostRun, user containers are expected to have completed (non-zero exit)

Reviewer: ```suggestion
- PostRun: Once in PostRun, user containers are expected to complete (non-zero exit)
```

successfully.
- DataOut: No container related activity.
- Teardown: Ports are released; `MPIJob` or `Job(s)` are deleted, which in turn deletes the user
containers.

The two main states of a container workflow (i.e. PreRun, PostRun) are discussed further in the
following sections.

### PreRun

> In PreRun, the containers are created and expected to start. Once the containers have reach a

Reviewer: ```suggestion
In PreRun, the containers are created and expected to start. Once the containers reach a
```

> non-initialization state (i.e. Running), the containers are considered to have been started and the

Reviewer: ```suggestion
non-initialization state (i.e. Running), the containers are considered to be started and the
```

workflow can advance.

By default, containers are expected to start within 60 seconds. If not, the workflow reports an
Error that the containers cannot be started. This value is configurable via the
`preRunTimeoutSeconds` field in the container profile.

To summarize the PreRun behavior:

- If the container starts successfully (running), transition to Ready:true.
- If the container fails to start, transition to the Error State.
- If the container is initializing and has not started after `preRunTimeoutSeconds` seconds,
terminate the container and transition to the Error State.

#### Init Containers

Containers have some additional initialization that occurs using [Init
Containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/).  These containers

> must run to completion before the main container can start. The NNF Software injects Init

Reviewer 2: This:
"The NNF Software injects Init Containers into the container specification"
is the key statement in this paragraph, and sub-section.  It should be the first sentence.

Author: I'll reword this

Containers into the container specification for various reasons:

- To ensure the proper permissions (i.e. user/group) are configured
- For MPI jobs, to ensure the launcher pod can contact the worker pods

### PostRun

> In PostRun, the containers are expected to exit cleanly with a zero exit code. If a container does

Reviewer: ```suggestion
In PostRun, the containers are expected to exit cleanly with a zero exit code. If a container fails to
```

> not exit cleanly, the Kubernetes software will attempt a number of retries based on the

Reviewer: ```suggestion
exit cleanly, the Kubernetes software attempts a number of retries based on the
```

> configuration of the container profile. It will continue to do this until the container exits

Reviewer: ```suggestion
configuration of the container profile. It continues to do this until the container exits
```

> successfully, or if the `retryLimit` is set - whichever occurs first. In the latter case, the

Reviewer: ```suggestion
successfully, or until the `retryLimit` is hit - whichever occurs first. In the latter case, the
```

> workflow will report an Error.

Reviewer: ```suggestion
workflow reports an Error.
```

Read up on the [Failure Retries](#failure-retries) for more information on retries.

Furthermore, the container profile features a `postRunTimeoutSeconds` field. If this timeout is
reached before the container successfully exits, it triggers an Error state. The timer for this
timeout begins upon entry into the PostRun phase, allowing the containers the specified period to
execute before the workflow enters an Error state.

To recap the PostRun behavior:

- If the container exits successfully, transition to Ready:true.
- If the container exits unsuccessfully after `retryLimit` number of retries, transition to the
Error State.
- If the container is running and has not exited after `postRunTimeoutSeconds` seconds, terminate
the container and transition to the Error State.

### Failure Retries

If a container fails (non-zero exit code), the Kubernetes software implements retries. The number of
retries can be set via the `retryLimit` field in the container profile. If a non-zero exit code is

> detected, the Kubernetes software will create a new instance of the pod and try again. The

Reviewer: ```suggestion
detected, the Kubernetes software creates a new instance of the pod and retries. The
```

default number of retries for `retryLimit` is set to 6, which is the default value for Kubernetes
Jobs. This means that if the pods fails every single time, there will be 7 failed pods in total
since it attempted 6 retries after the first failure.

To understand this behavior more, see [Pod backoff failure
policy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-backoff-failure-policy)
in the Kubernetes documentation. This explains the retry (i.e. backoff) behavior in more detail.

It is important to note that due to the configuration of the `MPIJob` and/or `Job` that is created
for User Containers, the container retries are immediate - there is no backoff timeout between
retires. This is due to the NNF Software setting the `RestartPolicy` to `Never`, which causes a new
pod to spin up after every failure rather than re-use (i.e. restart) the previously failed pod. This
allows a user to see a complete history of the failed pod(s) and the logs can easily be obtained.
See more on this at [Handling Pod and container
failures](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
in the Kubernetes documentation.

> ## Putting it All Together

Reviewer 2: I'm missing a piece: How does the user container know when PostRun begins?  Where does it get that cue?

Author: Is this regarding the PostRun timeout? The container has no idea - the NNF software sets the Job's activeDeadline when the workflow transitions to PostRun. Once that activeDeadline timeout is hit, k8s takes care of the rest and the containers are killed. The job transitions to a failed state and the NNF software turns that into a workflow error.

Reviewer 2: Is the user container supposed to do some work during PostRun?

Author: That's up to the application developer that created the image, but they won't have any cue from the NNF software that it's in PostRun.

I envision the port being the mechanism that drives behavior to the application (e.g. SCR) running in the container. The compute node application contacts the application in the container and tells it that it's time to do something.

1. In PreRun, user container starts successfully. The workflow transitions to PreRun complete
2. Flux runs the application on the compute node
3. The compute node application contacts the user container via the port to drive behavior, ultimately saying: "I'm done"
4. Compute node application finishes and Flux sees that
5. Flux transitions to PostRun
6. PostRun expects the user container to exit cleanly before transitioning to PostRun complete

--- 

@roehrich-hpe and I discussed this at my desk already, but posting this for posterity.

I'll also be adding a section that illustrates this more directly.

workflow can advance.

By default, containers are expected to start within 60 seconds. If not, the workflow reports an
Error that the containers cannot be started. This value is configurable via the
`preRunTimeoutSeconds` field in the container profile.

To summarize the PreRun behavior:

- If the container starts successfully (running), transition to `Completed` status.
- If the container fails to start, transition to the `Error` status.
- If the container is initializing and has not started after `preRunTimeoutSeconds` seconds,
terminate the container and transition to the `Error` status.

#### Init Containers

The NNF Software injects [Init
Containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) into the container
specification to perform initialization tasks. These containers must run to completion before the
main container can start.

These initialization tasks include:

- Ensuring the proper permissions (i.e. UID/GID) are available in the main container
- For MPI jobs, ensuring the launcher pod can contact each worker pod via DNS

### PreRun Completed

Once PreRun has transitioned to `Completed` status, the user container is now running.

> Sequentially, the WLM should initiate applications on the compute nodes. Utilizing [container

Reviewer 3: What do you mean by "sequentially" here? The user apps on the compute nodes should be able to start at the same time.

Author: I was making the assumption that Flux would wait for Completed before starting the application on the compute nodes. Maybe I should reword?

Reviewer 3: I read it as "start the application on compute nodes one at a time." Maybe that's just me.

Author: I can see that. I'll reword this.
